editor/translations/translation_catalog_maker.py

```diff
@@ -0,0 +1,237 @@
+        element._end_column_number = self.parser.CurrentColumnNumber
+        element._end_byte_index = self.parser.CurrentByteIndex
+        return element
+## </xml-line-number-hack>
+
+class Desc:
+	def __init__(self, line_no, msg, desc_list=None):
+		## line_no : the line number where the desc is
+		self.line_no = line_no
+		self.msg = msg
+		self.desc_list = desc_list
+
+class DescList:
+	def __init__(self, doc, path):
+		## doc  : root xml element of the document
+		## path : file path of the xml document
+		## list : list of Desc objects for this document
+		self.doc = doc
+		self.path = path
+		self.list = []
+
+def print_error(error):
+	print("ERROR: {}".format(error))
+    
+## build classes with xml elements recursively
+def _collect_classes_dir(path, classes):
+	if not os.path.isdir(path):
+		print_error("Invalid directory path :{}".format(path))
+		exit(1)
+	for _dir in map(lambda dir : os.path.join(path, dir), os.listdir(path)):
+		if os.path.isdir(_dir):
+			_collect_classes_dir(_dir, classes)
+		elif os.path.isfile(_dir):
+			if not _dir.endswith(".xml"):
+				#print("Got non-.xml file '{}', skipping.".format(path))
+				continue
+			_collect_classes_file(_dir, classes)
+
+## opens a file and parse xml add to classes
+def _collect_classes_file(path, classes):
+	if not os.path.isfile(path) or not path.endswith(".xml"):
+		print_error("Invalid xml file path :{}".format(path))
+		exit(1)
+	print('Collecting file: {}'.format(os.path.basename(path)))
+
+	try:
+		tree = ET.parse(path, parser=LineNumberingParser())
+	except ET.ParseError as e:
+		print_error("Parse error reading file '{}': {}".format(path, e))
+		exit(1)
+
+	doc = tree.getroot()
+
+	if 'version' not in doc.attrib:
+		print_error("Version missing from 'doc', file: {}".format(path))
+
+	if 'name' in doc.attrib:
+		name = doc.attrib["name"]
+		if name in classes:
+			print_error("Duplicate class {}:{}".format(path, name))
+			exit(1)
+		classes[name] = DescList(doc, path)
+	else:
+		print_error('Unknown XML file {}, skiping'.format(path))
+
+
+## regions are list of tuples with size 3 (start_index, end_index, indent)
+## indication in string where the codeblock starts, ends, and it's indent
+## if i inside the region returns the indent, else returns -1
+def _get_xml_indent(i, regions):
+	for region in regions:
+		if region[0] < i < region[1] :
+			return region[2]
+	return -1
+
+## find and build all regions of codeblock which we need later
+def _make_codeblock_regions(desc, path=''):
+	code_block_end = False
+	code_block_index = 0
+	code_block_regions = []
+	while not code_block_end:
+		code_block_index = desc.find("[codeblock]", code_block_index)
+		if code_block_index < 0: break
+		xml_indent=0
+		while True :
+			## [codeblock] always have a trailing new line and some tabs
+			## those tabs are belongs to xml indentations not code indent
+			if desc[ code_block_index + len("[codeblock]\n")+xml_indent] == '\t':
```

Comment:
Be sure to follow PEP-8 for formatting around `+` (spaces on each side) or within `[ ]` (no spaces).

We don't use an auto-formatter *yet*, but we'll likely use https://github.com/psf/black with a max line length of 120 (see #36566), so feel free to use it for this file.